### PR TITLE
Bridge actor

### PR DESCRIPTION
**Pull request by d-plaindoux**
Add bridge actor:

- [X] Clustering definition added
- [X] Region definition added
- [X] Cluster management added
- [X] Protocol reviewed between Region, Arbitrator, and Player
- [X] Bridge Actor definition

Starting the game:
```sh
$ sbt
> run estasia
... (Ugly port number not found / to be fixed later)
> run estasia
```

```sh
$ sbt
> run eurasia
... (Ugly port number not found / to be fixed later)
> run eurasia
```

### CODE REVIEW

File: src/main/scala/com/agar/core/Agar.scala [hunk 1]
```diff
@@ -1,24 +1,32 @@
...
-  val logger: ActorRef = system.actorOf(Logger.props, "logger")
-  val arbitrator: ActorRef = system.actorOf(Arbitrator.props(logger),"arbitrator")
+  val journal = system.actorOf(Logger.props, "logger")
+  val bridge = system.actorOf(Bridge.props(remotePort))
+  val region = system.actorOf(Region.props(bridge, journal, 7680, 4320), "region")
+  val arbitrator: ActorRef = system.actorOf(Arbitrator.props(bridge, region), "arbitrator")
 
-  // A region has a size of 4 screen 1920x1080
-  val region = system.actorOf(Region.props(arbitrator, logger, 7680, 4320), "region")
+  region ! InitRegion(nbPlayer, nbEnergy)
```
Comment (NotBad4U) [inline src/main/scala/com/agar/core/Agar.scala]: Maybe we should set a default value for this ?

**Comment by NotBad4U** [comment]
Do we have a way to start the project in single node only just for the test ?

**Comment by d-plaindoux** [comment]
The region definition (limited square) should be reviewed.

File: src/main/scala/com/agar/core/context/DefaultAgarSystem.scala [hunk 2]
```diff
@@ -4,6 +4,6 @@ import scala.concurrent.duration._
 
 object DefaultAgarSystem extends AgarSystem {
 
-  override def timeout(): FiniteDuration = 100 millis
+  override def timeout(): FiniteDuration = 1 second
```
Comment (NotBad4U) [inline src/main/scala/com/agar/core/context/DefaultAgarSystem.scala]: That'll be very slow and the client will not have a great refresh rate.

File: src/main/scala/com/agar/core/region/Region.scala [hunk 3]
```diff
@@ -1,88 +1,253 @@
...
+  private def createNewPlayer(position: Vector2d, weight: Int): ActorRef = {
+    context.actorOf(Player.props(position, weight)(self))
   }
 
   private def createNewEnergy(valueOfEnergy: Int): ActorRef = {
-    context.actorOf(Energy.props(valueOfEnergy))
+    context.actorOf(Energy.props(valueOfEnergy)(self))
+  }
+
+  private def isInRegion(position: Vector2d): Boolean = {
```
Comment (NotBad4U) [inline src/main/scala/com/agar/core/region/Region.scala]: This is not the correct to do this.
We should check whether a given circle (player) lies inside a square/rectangle or not.
https://yal.cc/rectangle-circle-intersection-test/
https://stackoverflow.com/a/1879223/5130134

File: src/main/scala/com/agar/core/region/Region.scala [hunk 4]
```diff
@@ -1,88 +1,253 @@
...
+    height / Math.abs(height) * position.y > -frontier
   }
+
+  private def isInFrontier(position: Vector2d): Boolean = {
```
Comment (NotBad4U) [inline src/main/scala/com/agar/core/region/Region.scala]: This is not the correct to do this.
But it's okai for the demo just create an issue related to this or add a `TODO`

File: src/test/scala/com/agar/core/region/RegionTest.scala [hunk 5]
```diff
@@ -23,25 +27,243 @@ class RegionTest(_system: ActorSystem)
...
+      expectedCorrectInit should be(true)
+
+      bridge.expectMsgPF(500 millis) {
+        case Virtual(RegisterPlayer(_, _)) => true
+      }
+
+    }
+  }
+
+  "move a player in the frontier" in {
```
Comment (NotBad4U) [inline src/test/scala/com/agar/core/region/RegionTest.scala]: I think we have a missing test: A player enter in the region
This test should check if a player is `create | virtual`

File: src/test/scala/com/agar/core/region/RegionTest.scala [hunk 6]
```diff
@@ -23,25 +27,243 @@ class RegionTest(_system: ActorSystem)
...
+
+    expectedCorrectInit should be(true)
+
+    bridge.expectMsgPF(500 millis) {
+      case Virtual(RegisterEnergy(_, _)) => true
+    }
+
+  }
+
+  "consume an energy in a frontier" in {
```
Comment (NotBad4U) [inline src/test/scala/com/agar/core/region/RegionTest.scala]: I think we are missing a test: kill a player in the region.
This is not an important test but we have the "kill a player in the frontier" so for me we should have the "kill a player in the region" too.

File: src/main/scala/com/agar/core/context/DefaultAgarSystem.scala (hunk 2 above)
Comment (d-plaindoux) [reply src/main/scala/com/agar/core/context/DefaultAgarSystem.scala]: Agree I change it

File: src/main/scala/com/agar/core/region/Region.scala (hunk 3 above)
Comment (d-plaindoux) [reply src/main/scala/com/agar/core/region/Region.scala]: Lake of time in fact. This issue has already been created.

File: src/main/scala/com/agar/core/region/Region.scala (hunk 4 above)
Comment (d-plaindoux) [reply src/main/scala/com/agar/core/region/Region.scala]: Lake of time in fact. This issue has already been created.

File: src/test/scala/com/agar/core/region/RegionTest.scala (hunk 6 above)
Comment (d-plaindoux) [reply src/test/scala/com/agar/core/region/RegionTest.scala]: Tests added

File: src/test/scala/com/agar/core/region/RegionTest.scala (hunk 5 above)
Comment (d-plaindoux) [reply src/test/scala/com/agar/core/region/RegionTest.scala]: This is covered by "move a player out the frontier and the region"

File: src/main/scala/com/agar/core/Agar.scala (hunk 1 above)
Comment (d-plaindoux) [reply src/main/scala/com/agar/core/Agar.scala]: It's set up using the configuration file

File: src/main/scala/com/agar/core/context/DefaultAgarSystem.scala (hunk 2 above)
Comment (d-plaindoux) [reply src/main/scala/com/agar/core/context/DefaultAgarSystem.scala]: The definition is unused now. The current timeout used is set up to 200 milliseconds.

Comment (NotBad4U) [reply src/main/scala/com/agar/core/context/DefaultAgarSystem.scala]: Kep

File: src/main/scala/com/agar/core/region/Region.scala (hunk 3 above)
Comment (NotBad4U) [reply src/main/scala/com/agar/core/region/Region.scala]: Kep fine enough

File: src/test/scala/com/agar/core/region/RegionTest.scala (hunk 6 above)
Comment (NotBad4U) [reply src/test/scala/com/agar/core/region/RegionTest.scala]: thx !

File: src/test/scala/com/agar/core/region/RegionTest.scala (hunk 5 above)
Comment (NotBad4U) [reply src/test/scala/com/agar/core/region/RegionTest.scala]: Okay my bad

**Comment by NotBad4U** [comment]
Thx !

**Pull request merged**